zipline/pipeline/loaders/blaze.py

```diff
@@ -0,0 +1,870 @@
+
+
+from ..data.dataset import DataSet, Column
+from zipline.lib.adjusted_array import adjusted_array
+from zipline.lib.adjustment import Float64Overwrite
+
+
+AD_FIELD_NAME = 'asof_date'
+TS_FIELD_NAME = 'timestamp'
+SID_FIELD_NAME = 'sid'
+valid_deltas_node_types = (
+    bz.expr.Field,
+    bz.expr.ReLabel,
+    bz.expr.Symbol,
+)
+traversable_nodes = (
+    bz.expr.Field,
+    bz.expr.Label,
+)
+is_invalid_deltas_node = complement(flip(isinstance, valid_deltas_node_types))
+getname = op.attrgetter('__name__')
+
+
+class _ExprRepr(object):
+    """Box for repring expressions with the str of the expression.
+
+    Parameters
+    ----------
+    expr : Expr
+        The expression to box for repring.
+    """
+    __slots__ = 'expr',
+
+    def __init__(self, expr):
+        self.expr = expr
+
+    def __repr__(self):
+        return str(self.expr)
+    __str__ = __repr__
+
+
+class ExprData(namedtuple('ExprData', 'expr deltas resources')):
+    """A pair of expressions and a data resources.
+
+    Parameters
+    ----------
+    expr : Expr
+        The first known values.
+    deltas : Expr, optional
+        The deltas for the data.
+    resources : resource or dict of resources, optional
+        The resources to compute the exprs against.
+    """
+    def __new__(cls, expr, deltas=None, resources=None):
+        return super(ExprData, cls).__new__(cls, expr, deltas, resources)
+
+    def __repr__(self):
+        # If the expressions have _resources() then the repr will
+        # drive computation so we box them.
+        cls = type(self)
+        return super(ExprData, cls).__repr__(cls(
+            _ExprRepr(self.expr),
+            _ExprRepr(self.deltas),
+            self.resources,
+        ))
+
+
+class InvalidField(with_metaclass(ABCMeta)):
+    """A field that raises an exception that indicates that the
+    field was invalid.
+
+    Parameters
+    ----------
+    field : str
+        The name of the field.
+    type_ : dshape
+        The shape of the field.
+    """
+    @abstractproperty
+    def error_format(self):
+        raise NotImplementedError('error_format')
+
+    def __init__(self, field, type_):
+        self._field = field
+        self._type = type_
+
+    def __get__(self, instance, owner):
+        raise AttributeError(
+            self.error_format.format(field=self._field, type_=self._type),
+        )
+
+
+class NonNumpyField(InvalidField):
+    error_format = "field '{field}' was a non numpy compatible type: '{type_}'"
+
+
+class NonPipelineField(InvalidField):
+    error_format = (
+        "field '{field}' was a non Pipeline API compatible type:"
+        " '{type_.__name__}'"
+    )
+
+
+class NotPipelineCompatible(TypeError):
+    """Exception used to indicate that a dshape is not Pipeline API
+    compatible.
+    """
+    def __str__(self):
+        return "'%s' is a non Pipleine API compatible type'" % self.args
+
+
+_new_names = ('_%d' % n for n in count())
```

Comment:
Can this be "BlazeDataSet_%d" so that people can easily tell where their weirdly-named dataset came from?